File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution():
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        cycleNode = None
        hare = head
        tortoise = head
        if hare == None:
          return None
        while True:
          hare = hare.next
          if hare == None:
            return None
          hare = hare.next
          if hare == None:
            return None
          tortoise = tortoise.next
          if hare == tortoise:
            cycleNode = hare
            break
        temp = head
        while temp != cycleNode:
          temp = temp.next
          cycleNode = cycleNode.next
        return cycleNode

File: 142-linked-list-cycle-ii/142-linked-list-cycle-ii.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
       map<ListNode*, bool> mp;
      while(true)
      {
        if(head==NULL)
        {
          return NULL;
        }
        if(mp[head])
        {
          return head;
        }
        else
        {
          mp[head] = true; 
        }
        head=head->next;
      }
      return NULL;
    }
};